File: pw_rpc/examples/DocsExample.java
// Copyright 2026 The Pigweed Authors
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.

package dev.pigweed.pw_rpc.examples;

import com.google.common.util.concurrent.FutureCallback;
import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.common.util.concurrent.MoreExecutors;
import dev.pigweed.pw_rpc.Call;
import dev.pigweed.pw_rpc.Channel;
import dev.pigweed.pw_rpc.Client;
import dev.pigweed.pw_rpc.MethodClient;
import dev.pigweed.pw_rpc.Service;
import dev.pigweed.pw_rpc.Status;
import dev.pigweed.pw_rpc.StreamObserver;
import dev.pigweed.pw_rpc.UnaryResult;
import java.nio.ByteBuffer;
import java.util.List;

/** Example snippets for pw_rpc Java client documentation. */
public final class DocsExample {
  private DocsExample() {}

  // DOCSTAG: [pw_rpc-java-create-client]
  public static Client createClient(Service sensorService) {
    // 1. Define channel output
    Channel channel = new Channel(1, (byte[] data) -> {
      // Transmit byte buffer over physical transport (USB, Bluetooth, etc.)
      System.out.println("Sending " + data.length + " bytes to device");
    });

    // 2. Instantiate Client
    return Client.createMultiCall(List.of(channel), List.of(sensorService));
  }
  // DOCSTAG: [pw_rpc-java-create-client]

  // DOCSTAG: [pw_rpc-java-route-packets]
  public static void onDataReceived(Client client, byte[] rawPacket) {
    client.processPacket(ByteBuffer.wrap(rawPacket));
  }
  // DOCSTAG: [pw_rpc-java-route-packets]

  // DOCSTAG: [pw_rpc-java-unary-call]
  public static <TReq, TResp> Call callUnary(
      Client client, MethodClient methodClient, TReq request) {
    return methodClient.invokeUnary(request, new StreamObserver<TResp>() {
      @Override
      public void onNext(TResp response) {
        System.out.println("Received response: " + response);
      }

      @Override
      public void onCompleted(Status status) {
        System.out.println("RPC finished with status: " + status);
      }

      @Override
      public void onError(Status status) {
        System.err.println("RPC failed with error: " + status);
      }
    });
  }
  // DOCSTAG: [pw_rpc-java-unary-call]

  // DOCSTAG: [pw_rpc-java-server-streaming-call]
  public static <TReq, TResp> Call callServerStreaming(
      Client client, MethodClient methodClient, TReq request) {
    return methodClient.invokeServerStreaming(request, new StreamObserver<TResp>() {
      @Override
      public void onNext(TResp response) {
        System.out.println("Stream response: " + response);
      }

      @Override
      public void onCompleted(Status status) {
        System.out.println("Stream completed: " + status);
      }

      @Override
      public void onError(Status status) {
        System.err.println("Stream error: " + status);
      }
    });
  }
  // DOCSTAG: [pw_rpc-java-server-streaming-call]

  // DOCSTAG: [pw_rpc-java-streaming-call]
  public static <TReq, TResp> void callBidirectionalStreaming(
      Client client, MethodClient methodClient, TReq chunk1, TReq chunk2) {
    StreamObserver<TResp> observer = new StreamObserver<>() {
      @Override
      public void onNext(TResp response) {
        System.out.println("Stream response: " + response);
      }

      @Override
      public void onCompleted(Status status) {
        System.out.println("Stream finished: " + status);
      }

      @Override
      public void onError(Status status) {
        System.err.println("Stream error: " + status);
      }
    };

    Call.ClientStreaming<TReq> stream = methodClient.invokeBidirectionalStreaming(observer);

    // Stream requests to server
    stream.send(chunk1);
    stream.send(chunk2);

    // Finish sending from client
    stream.finish();
  }
  // DOCSTAG: [pw_rpc-java-streaming-call]

  // DOCSTAG: [pw_rpc-java-future-call]
  public static <TReq, TResp> void callUnaryFuture(
      Client client, MethodClient methodClient, TReq request) {
    ListenableFuture<UnaryResult<TResp>> future = methodClient.invokeUnaryFuture(request);

    // Access result or attach listeners:
    Futures.addCallback(future, new FutureCallback<UnaryResult<TResp>>() {
      @Override
      public void onSuccess(UnaryResult<TResp> result) {
        if (result.status().ok()) {
          System.out.println("Result: " + result.response());
        }
      }

      @Override
      public void onFailure(Throwable t) {
        t.printStackTrace();
      }
    }, MoreExecutors.directExecutor());
  }
  // DOCSTAG: [pw_rpc-java-future-call]
}

File: pw_rpc/ts/docs_example.ts
// Copyright 2026 The Pigweed Authors
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.

import { Message } from 'google-protobuf';
import { Status } from 'pigweedjs/pw_status';
import { ProtoCollection } from 'pigweedjs/protos/collection';
import { RpcPacket } from 'pigweedjs/protos/pw_rpc/internal/packet_pb';
import {
  BidirectionalStreamingMethodStub,
  Channel,
  Client,
  ClientStreamingMethodStub,
  ServerStreamingMethodStub,
  UnaryMethodStub,
} from './index';

// DOCSTAG: [pw_rpc-ts-create-client]
function savePacket(packetBytes: Uint8Array): void {
  const packet = RpcPacket.deserializeBinary(packetBytes);
  // Forward packet to physical transport:
  // transport.send(packet.serializeBinary());
  console.log(`Sending packet to channel ${packet.getChannelId()}`);
}

const channels = [new Channel(1, savePacket), new Channel(5)];
const client = Client.fromProtoSet(channels, new ProtoCollection());
// DOCSTAG: [pw_rpc-ts-create-client]

// DOCSTAG: [pw_rpc-ts-find-method]
const channel = client.channel()!;
const unaryStub = channel.methodStub(
  'pw.rpc.test1.TheTestService.SomeUnary',
) as UnaryMethodStub;
// DOCSTAG: [pw_rpc-ts-find-method]

// DOCSTAG: [pw_rpc-ts-callback-invocation]
const bidiStub = client
  .channel()!
  .methodStub(
    'pw.rpc.test1.TheTestService.SomeBidi',
  ) as BidirectionalStreamingMethodStub;

// Configure callback functions:
const onNext = (response: Message) => {
  console.log(`Received message: ${response}`);
};
const onComplete = (status: Status) => {
  console.log(`RPC completed with status: ${status}`);
};
const onError = (error: Error) => {
  console.error(`RPC error: ${error}`);
};

const request = new bidiStub.method.requestType();
bidiStub.invoke(request, onNext, onComplete, onError);
// DOCSTAG: [pw_rpc-ts-callback-invocation]

// DOCSTAG: [pw_rpc-ts-unary-promise]
async function callUnary(): Promise<void> {
  const unaryRpc = client
    .channel()!
    .methodStub('pw.rpc.test1.TheTestService.SomeUnary') as UnaryMethodStub;

  const req = new unaryRpc.method.requestType();
  const timeoutMs = 2000;
  const [status, response] = await unaryRpc.call(req, timeoutMs);
  console.log(`Status: ${status}, Response: ${response}`);
}
// DOCSTAG: [pw_rpc-ts-unary-promise]

// DOCSTAG: [pw_rpc-ts-server-streaming-promise]
async function callServerStreaming(): Promise<void> {
  const serverStreamRpc = client
    .channel()!
    .methodStub(
      'pw.rpc.test1.TheTestService.SomeServerStreaming',
    ) as ServerStreamingMethodStub;

  const req = new serverStreamRpc.method.requestType();
  const call = serverStreamRpc.invoke(req);
  const timeoutMs = 2000;

  // Stream responses as they arrive:
  for await (const response of call.getResponses(2, timeoutMs)) {
    console.log(response);
  }

  // Await remaining responses until stream completion:
  const responses = call.getResponses();
  while (!responses.done) {
    console.log(await responses.value());
  }
}
// DOCSTAG: [pw_rpc-ts-server-streaming-promise]

// DOCSTAG: [pw_rpc-ts-client-streaming-promise]
async function callClientStreaming(): Promise<void> {
  const clientStreamRpc = client
    .channel()!
    .methodStub(
      'pw.rpc.test1.TheTestService.SomeClientStreaming',
    ) as ClientStreamingMethodStub;

  const stream = clientStreamRpc.invoke();
  const req = new clientStreamRpc.method.requestType();

  // Send request messages to the stream:
  stream.send(req);

  // Complete stream and await final unary response:
  const timeoutMs = 2000;
  stream
    .finishAndWait([req, req], timeoutMs)
    .then(([status, response]) => {
      console.log(`Finished: ${status}, Response: ${response}`);
    })
    .catch((reason) => {
      console.error(`Stream error: ${reason}`);
    });
}
// DOCSTAG: [pw_rpc-ts-client-streaming-promise]

// DOCSTAG: [pw_rpc-ts-bidi-streaming-promise]
async function callBidiStreaming(): Promise<void> {
  const bidiStreamingRpc = client
    .channel()!
    .methodStub(
      'pw.rpc.test1.TheTestService.SomeBidiStreaming',
    ) as BidirectionalStreamingMethodStub;

  const stream = bidiStreamingRpc.invoke();
  const req = new bidiStreamingRpc.method.requestType();

  // Send requests to device:
  stream.send(req);

  // Receive stream responses:
  const timeoutMs = 2000;
  for await (const response of stream.getResponses(1, timeoutMs)) {
    console.log(response);
  }

  // Finish sending and await completion:
  stream
    .finishAndWait([req], timeoutMs)
    .then(([status]) => {
      console.log(`Bidirectional stream finished with status: ${status}`);
    })
    .catch((reason) => {
      console.error(`Bidirectional stream error: ${reason}`);
    });
}
// DOCSTAG: [pw_rpc-ts-bidi-streaming-promise]
